src/test/test-alloc-util.c

```diff
@@ -21,21 +22,35 @@ static void test_alloca(void) {
 
 static void test_GREEDY_REALLOC(void) {
         _cleanup_free_ int *a = NULL, *b = NULL;
-        size_t n_allocated = 0, i;
+        size_t n_allocated = 0, i, j;
 
-        /* Give valgrind a chance to verify our realloc operations */
+        /* Give valgrind a chance to verify our realloc() operations */
 
-        for (i = 0; i < 2048; i++) {
+        for (i = 0; i < 20480; i++) {
                 assert_se(GREEDY_REALLOC(a, n_allocated, i + 1));
-                a[i] = i;
+                assert_se(n_allocated >= i + 1);
+                assert_se(malloc_usable_size(a) >= (i + 1) * sizeof(int));
+                a[i] = (int) i;
                 assert_se(GREEDY_REALLOC(a, n_allocated, i / 2));
+                assert_se(n_allocated >= i / 2);
+                assert_se(malloc_usable_size(a) >= (i / 2) * sizeof(int));
         }
 
-        for (i = 30, n_allocated = 0; i < 2048; i+=7) {
+        for (j = 0; j < i / 2; j++)
```

Comment:
This should loop over all elements rather than 1/2 of them for the check. I suspect that this was done as an optimization, but there's two reasons why I would advice against this and instead iterate over all valid elements:

1. The current number of elements in the array is 20480, but since this isn't a power of two I would expect the allocated size to be 32768 elements large. As a result, looping over i/2 elements only checks 10240 of the 16384 elements that have been copied over by the last reallocation, so it isn't fully checking that all elements have been preserved.
2. This is an optimization based on the current implementation of the function. Since a large part of unit tests is to catch regressions if the implementation of a function changes, it would be safer to be more thorough with the checks even if it's a bit slower.

Reply:
well, we called GREEDY_REALLOC earlier with only half the space, hence checking more than half is kinda accessing memory we don't really own anymore

Reply:
Sorry, my mistake.